[1단계 - 로또 구현] 레오(김준영) 미션 제출합니다.

kjyyjk opened the pull request:
안녕하세요 로키~
7기 레오입니다!

처음 접한 페어 프로그래밍이라 여러모로 걱정되기도 했지만,
페어인 듀이와 함께 의견을 조율하고 하나의 결과물을 만들어가며 즐거운 시간이 되었어요.

학습 로그 : https://prolog.techcourse.co.kr/studylogs/3834

추가로 아래 사항에 대한 로키의 의견이 궁금하여 남겨드려요.
1. static 키워드 사용
  LottoManager, LottoMachine의 public 메서드들을 static으로 선언하여 사용했습니다. 여러 메서드에서 매번 인스턴스를 생성하기 보다는, LottoManager와 LottoMachine 모두 상태를 가지지 않기 때문에 static을 사용하면 좋을 것이라고 판단했습니다.

2. Application main에서 프로그램의 흐름을 제어
  현재는 프로그램의 흐름에 대한 로직이 Application의 main 메서드에 존재합니다. MVC 패턴의 Controller와 같이 별도의 클래스에서 애플리케이션 흐름을 관리해야 할 것인가에 대해 페어와 함께 의논을 해보았으나, 현재 단계에서는 큰 이유를 찾지 못하였기 때문에 변경을 하지는 않았습니다.

3. private method 내부 로직 테스트
주석으로 남겨둔 TODO를 commit하여 기록해두기도 했습니다만, LottoMachine 클래스의 calculate 메서드 내에 아래와 같은 로직이 존재합니다.
```
private static Prize calculate(final List<Integer> lottoNumbers, final List<Integer> winningNumbers,
                                   final int bonusNumber) {
        List<Integer> matchNumbers = new ArrayList<>(winningNumbers);
        matchNumbers.retainAll(lottoNumbers);
        ...
}
```
이때 두 리스트의 교집합을 올바르게 반환하는지 테스트하고 싶은데, private method 내부의 로직이다 보니 해당 메서드를 테스트할 수가 없었습니다. 따라서 해당 부분만 별도의 메서드로 분리하여 테스트해볼까 했지만 과연 테스트만을 위해서 메서드를 만들어도 되는지, 혹은 테스트가 보내는 잘못된 설계의 신호라고 생각하고 진행해야할 지와 같은 고민들을 하게 되어 TODO로 남겨두었습니다.

감사합니다!

kjyyjk committed:
docs: 기능 목록 작성

kjyyjk committed:
docs: 기능 목록 작성

kjyyjk committed:
docs: 기능 목록 수정

- 로또의 발행 방법을 구체적으로 명시한다.

kjyyjk committed:
feat: 로또 발행

- 로또는 1 ~ 45 사이 6개 숫자를 랜덤으로 생성하여 발행한다.

- 로또 번호는 중복을 허용하지 않기 때문에 Set을 사용하였다.

- 난수 발생에 따라 테스트의 멱등성이 깨질 수 있기 때문에 리팩토링이 필요하다.

ljhee92 committed:
feat: 구입 금액 입력

ljhee92 committed:
test: 구입 금액 검증
- 구입 금액이 1000원으로 나누어지지 않으면 IllegalArgumentException을 던진다.
- 구입 금액이 양수가 아니라면 IllegalArgumentException을 던진다.

ljhee92 committed:
feat: 구입 금액에 따른 로또 발행

ljhee92 committed:
feat: 구입 금액 입력 검증

kjyyjk committed:
feat: 발행된 로또 출력

- Set은 정렬을 지원하지 않기 때문에 LottoMachine에서
정렬된 List를 반환한다.

kjyyjk committed:
refactor: Lotto 객체화

- 로또가 핵심 도메인이기도 하고 List 내 List가 거슬려서
Lotto를 따로 객체화하였다.

- 실제 로또도 발행시 정렬되어 나오기 때문에 Lotto 생성 시
생성자 내부에서 정렬하게끔 구현했다.

ljhee92 committed:
feat: 최대 구입 금액 검증
- 구입 금액이 10만원을 넘으면 IllegalArgumentException을 던진다.

ljhee92 committed:
feat: 당첨 번호 입력

ljhee92 committed:
feat: 보너스 번호 입력

ljhee92 committed:
refactor: 예외 발생 시 입력값 재입력
- 실제 로또를 생각해 봤을 때 잘못된 입력에 예외를 던지고 프로그램을 종료하기보다
안내 메시지를 출력하고 다시 입력을 받는 것이 적절하다고 생각하여 리팩토링했다.

ljhee92 committed:
docs: 당첨 번호, 보너스 번호 예외 추가

ljhee92 committed:
feat: 당첨 번호 검증
- 입력한 당첨 번호가 1 ~ 45 사이가 아니면 IllegalArgumentException을 발생시키고, 재입력 받는다.
- 입력한 당첨 번호가 고유한 6개의 번호가 아니라면 IllegalArgumentException을 발생시키고, 재입력 받는다.

kjyyjk committed:
feat: 보너스 번호 검증

- 입력한 보너스 번호가 1 ~ 45 사이가 아니면 IllegalArgumentException을 발생시키고, 재입력 받는다.

kjyyjk committed:
feat: 보너스 번호 중복 검증

- 입력한 보너스 번호가 당첨 번호와 중복된다면IllegalArgumentException을 발생시키고,
재입력 받는다.

- ,를 기준으로 자른뒤 공백을 없애주기 위해 trim을 사용한다.

kjyyjk committed:
feat: 당첨 개수 계산

- LottoMachine과 LottoManager의 메서드를 static으로 변경하였다.
여러곳에서 사용되지만 매번 인스턴스화하기는 낭비라고 판단했고, 싱글턴을 적용할 이유는 찾지
못하였다. 그리고 두 클래스가 상태를 가지지 않기 때문에 static을 사용하였다.

- 당첨 금액과 번호 일치 개수, 보너스 번호 일치 여부를 가지는 Prize enum 상수를 사용했다.

ljhee92 committed:
feat: 당첨 수익률 계산
- 수익률 계산을 위해 Map을 가진 WinningStatistics 클래스를 생성하였다.

ljhee92 committed:
feat: 당첨 통계 출력

ljhee92 committed:
refactor: import 정리

ljhee92 committed:
refactor: view 분리

ljhee92 committed:
chore: gitkeep 파일 삭제

kjyyjk committed:
refactor: 매개변수 final 추가

kjyyjk committed:
refactor: Prize.getPrize() 내부 로직 변경

- 들여쓰기 단계를 최소화하기 위해 변경하였다.

kjyyjk committed:
refactor: 매직넘버 상수처리

kjyyjk committed:
refactor: 로또 번호 검증로직 위치변경

- Lotto.validateLottoNumber 단위 테스트

kjyyjk committed:
refactor: 매직넘버 상수 처리

kjyyjk committed:
refactor: 테스트에서만 사용되는 메서드 제거

ljhee92 committed:
docs: 테스트에 대한 고민 TODO

ljhee92 committed:
refactor: parseInt 공통 예외처리

Rok93 commented on README.md:
```diff
@@ -1,3 +1,31 @@
+- [x] 당첨 통계를 출력한다
+
+## 예외
+- [x] 구입 금액이 숫자가 아니면 IllegalArgumentException을 발생시키고, 재입력 받는다.
+- [x] 구입 금액이 1000원으로 나누어지지 않으면 IllegalArgumentException을 발생시키고, 재입력 받는다.
+- [x] 구입 금액이 양수가 아니라면 IllegalArgumentException을 발생시키고, 재입력 받는다.
+- [x] 구입 금액이 10만원을 넘으면 IllegalArgumentException을 발생시키고, 재입력 받는다.
+- [x] 입력한 당첨 번호가 1 ~ 45 사이가 아니면 IllegalArgumentException을 발생시키고, 재입력 받는다.
+- [x] 입력한 보너스 번호가 1 ~ 45 사이가 아니면 IllegalArgumentException을 발생시키고, 재입력 받는다.
+- [x] 입력한 당첨 번호가 고유한 6개의 번호가 아니라면 IllegalArgumentException을 발생시키고, 재입력 받는다.
+- [x] 입력한 보너스 번호가 당첨 번호와 중복된다면 IllegalArgumentException을 발생시키고, 재입력 받는다.
```
예외 케이스는 이렇게 따로 빼서 관리하기보다는 
기능 요구사항 목록의 하위 뎁스로 예외들을 추가해주면 좋을 것 같아요! 😃

kjyyjk replied:
반영했습니다!

Rok93 commented on README.md:
```diff
@@ -1,3 +1,31 @@
+- 입력값이 올바르지 않으면 안내 메시지를 출력하고 재입력 받는다.
+- 총 수익률이 1 미만일 때만 손해임을 안내한다.
+
+## 기능 목록
```
기능요구사항 잘 작성해주셨네요 👍💯

Rok93 commented on src/main/java/lotto/Application.java:
```diff
@@ -0,0 +1,74 @@
+import static lotto.Lotto.LOTTO_SIZE;
+import static lotto.Lotto.validateLottoNumber;
+
+import lotto.view.InputView;
```
PR 본문으로 질문주신 내용들은 여기에서 모아서 답변드릴게요. 😃

> 질문1
static 키워드 사용
LottoManager, LottoMachine의 public 메서드들을 static으로 선언하여 사용했습니다. 여러 메서드에서 매번 인스턴스를 생성하기 보다는, LottoManager와 LottoMachine 모두 상태를 가지지 않기 때문에 static을 사용하면 좋을 것이라고 판단했습니다.

말씀하신 것처럼 상태를 가질 필요가 없다면 객체를 생성할 필요가 없으니, public static 메서드를 두는 것이 자연스러운 흐름이라 생각합니다.  다만 두 가지 포인트로 고려해야할 것이 있을 것 같은데요. 두 부분을 고려해보시면 좋을 것 같습니다. 

1. 먼저 정말로 상태가 필요없는지 고민해볼 수 있을 것 같아요.  (지금 설계가 잘못됐다는 것이 아니라 한번쯤 다시 점검해볼 필요가 있다 정도로 생각해주세요. 😉)
2. 불필요한 객체 생성을 방지하는 조치 (기본생성자의 접근제한자를 private으로 설정) 

(이 코멘트를 써두고 코드를 훑어보고있었는데, 2번의 처리는 이미 잘해주셨군요? 👍💯)

개인적인 생각으로는 LottoMachine 혹은 LottoManager가 돈을 받아서 로또를 발권해주는 것도 괜찮다고 생각해요. 
아마 레오는 로또 판매점에 가서 로또 판매점 직원에게 돈을 내고 복권을 사는 상황을 생각하면서 모델링을 하신 것 같아요. 😃
저는 자판기처럼 돈을 지불하면 로또 자판기(= 로또머신)이 지불한 돈만큼 자동으로 복권을 지급해주는 상황을 머릿속으로 그려볼 수도 있을 것 같아요. 

> 질문2
Application main에서 프로그램의 흐름을 제어
현재는 프로그램의 흐름에 대한 로직이 Application의 main 메서드에 존재합니다. MVC 패턴의 Controller와 같이 별도의 클래스에서 애플리케이션 흐름을 관리해야 할 것인가에 대해 페어와 함께 의논을 해보았으나, 현재 단계에서는 큰 이유를 찾지 못하였기 때문에 변경을 하지는 않았습니다.

다음 단계 요구사항을 몰래 살짝 보면 레오가 언급하신 MVC패턴, Controller을 접목시켜 리팩터링을해야하실텐데요. 
2단계에서 이런 얘기를 상황에 따라 더 해보면 좋을 것 같습니다. 

Controller는 애플리케이션의 흐름을 담당하는 역할이 아니라는 정도만 맛보기로 말씀드릴게요. 😃
필요성을 못느끼셨다면 지금의 상태를 유지해도 괜찮을 것 같습니다. 😉 


> 질문3
private method 내부 로직 테스트
주석으로 남겨둔 TODO를 commit하여 기록해두기도 했습니다만, LottoMachine 클래스의 calculate 메서드 내에 아래와 같은 로직이 존재합니다.
...
이때 두 리스트의 교집합을 올바르게 반환하는지 테스트하고 싶은데, private method 내부의 로직이다 보니 해당 메서드를 테스트할 수가 없었습니다. 따라서 해당 부분만 별도의 메서드로 분리하여 테스트해볼까 했지만 과연 테스트만을 위해서 메서드를 만들어도 되는지, 혹은 테스트가 보내는 잘못된 설계의 신호라고 생각하고 진행해야할 지와 같은 고민들을 하게 되어 TODO로 남겨두었습니다.

먼저 이 LottoMachine의 (private)한 메서드를 테스트하고 싶으신 이유가 무엇인지 여쭙고싶어요. 😃
아마 추측컨테 중요한 로직이라고 생각되셨기 때문일거라고 생각해요. 

저도 한 때, 이런 고민을 한 적이 있엇고, 지금까지도 늘 기억에 남는 답변이 있는데요. 
그 답변을 대신해서 공유드려봅니다. 😃

> private method에 대한 단위 테스트가 필요하다고 느끼는 순간 무엇인가 bad smell이 있는 것은 아닐까라는 고민을 할 수 있는 시점이라고 봅니다. 즉, sum method가 Statistics 클래스에 위치하는 것이 맞지 않고 별도의 클래스로 이동해야 하는 것이 아닐까라는 고민을 할 수 있겠죠. Sum이라는 클래스를 추가할 수도 있는지를 검토해 볼 수 있는 것이겠죠. method의 위치를 이동하는 순간 private이던 method가 public으로 변경될 가능성도 높아질 겁니다. 그 시점에 단위 테스트를 추가할 수 있겠죠. 

[참조](https://www.slipp.net/questions/253#answer-983)

kjyyjk replied:
답변 감사합니다!! 꼭 참고할 수 있도록 해볼께요😃

Rok93 commented on src/main/java/lotto/Application.java:
```diff
@@ -0,0 +1,74 @@
+
+    private static int getBonusNumber(final WinningNumbers winningNumbers) {
+        try {
+            int bonusNumber = InputView.inputBonusNumber();
+            validateLottoNumber(bonusNumber);
+            winningNumbers.validateBonusNumberDuplicated(bonusNumber);
```
이렇게 관리하기보다는  당첨번호들과, 보너스번호를 상태로 가지는 `당첨 로또` 객체를 만들어보면 어떨까요? 
그러면 당첨 로또 객체가 생성되는 시점에, `validateBonusNumberDuplicated` 검증도 수행할 수 있을 것 같아요. 😃

kjyyjk replied:
로키의 조언대로 WinningLotto를 만들어서 중복 검증 로직을 WinningLotto 내부로 이동했습니다~!
추가로 당첨 번호 검증 로직도 WinningNumbers로 이동해 생성자 내에서 호출하도록 수정하였습니다.
그 과정에서 WinningNumbers 생성 단위 테스트를 진행하고 처음으로 MethodSource를 사용해볼 수 있었습니다😃

kjyyjk replied:
추가로 당첨 번호 검증 로직 또한 WinningNumbers로 이동해 생성자 내에서 호출하도록 수정하였습니다.

그 과정에서 WinningNumbers 생성 단위 테스트를 진행할 수 있게 되었고,
처음으로 MethodSource를 활용한 Parameterized Test도 사용해보았습니다!

Rok93 commented on src/main/java/lotto/Lotto.java:
```diff
@@ -0,0 +1,37 @@
+public class Lotto {
+    public static final int MAX_LOTTO_NUMBER = 45;
+    public static final int MIN_LOTTO_NUMBER = 1;
+    public static final int LOTTO_SIZE = 6;
+    private final List<Integer> numbers;
```
상수와 상태는 줄바꿈을 통해서 구분해주시면 가독성이 더 좋아질 것 같아요. 😉

kjyyjk replied:
확인하고 수정했습니다!
같은 이유로 OutputView에서 상수와 생성자 간에도 줄바꿈 추가했습니다😃

kjyyjk replied:
반영했습니다😃

Rok93 commented on src/main/java/lotto/Lotto.java:
```diff
@@ -0,0 +1,37 @@
+        }
+    }
+
+    public List<Integer> getNumbers() {
```
방어적 복사 👍💯

kjyyjk replied:
이런 방식을 방어적 복사라고 하는군요!
하나 궁금한 것이 있는데 numbers.stream().toList()를 반환해도 방어적 복사가 되는 것으로 알고 있습니다.
실제로 WinningNumbers 클래스의 getWinningNumbers()에서는 해당 방식을 사용하고 있는데,
로키는 보통 어떤 방식을 사용하시나요?
어떤 것을 사용해도 괜찮다면 한가지 방식으로 통일해주는 것이 좋을까요?

Rok93 replied:
stream()을 호출하게되면 컬렉션을 먼저 Stream으로 변환하게되고 그 이후에 toList를 호출할 때, List를 호출한다고 생각해요. 
방식이 뭐가됐건 방어적 복사가 된다면 상관없지만, 레오가 질문주신 방식은 불필요한 변환이 끼어있다고 생각해요. 
불필요한 변환을 없애는 것이 낫다고 생각되는 편이에요. 😃 

방어적 복사에 대해서는 [이 글](https://tecoble.techcourse.co.kr/post/2021-04-26-defensive-copy-vs-unmodifiable/)을 조금 더 참고해보시면 좋을 것 같습니다. 😃

Rok93 commented on src/main/java/lotto/LottoMachine.java:
```diff
@@ -0,0 +1,49 @@
+        return new Lotto(numbers);
+    }
+
+    public static WinningStatistics calculateStatistics(final List<Lotto> lottos, final WinningNumbers winningNumbers,
```
`당첨 통계를 계산해주는 역할`을 로또 머신 대신에 다른 객체에게 위임해보면 어떨까요? 😃
이렇게 역할을 위임하기 위해서 새로운 객체를 만들어보시면 좋을 것 같은데요. 

현재 메서드 파라미터로 받고있는 값들 혹은 완전히 새로운 객체를 생성할 수도 있을 것 같아요. 😃

kjyyjk replied:
새로 생성한 WinningLotto 객체에서 당첨 통계를 계산하는게 적절하다고 판단되어
WinningLotto에게 위임하였습니다!

Rok93 commented on src/main/java/lotto/Prize.java:
```diff
@@ -0,0 +1,49 @@
+package lotto;
+
+public enum Prize {
```
개인적인 생각이지만 등수를 표현하신거라면 Prize보다는 Rank 같은게 조금 더 직관적이지 않을까요? 👀

kjyyjk replied:
네 Rank가 더 직관적일 것이라고 생각되어 반영했습니다!

Rok93 commented on src/main/java/lotto/Prize.java:
```diff
@@ -0,0 +1,49 @@
+    }
+
+    public static Prize getPrize(final int matchCount, final boolean hasBonusNumber) {
+        if (matchCount == FIRST.matchCount) {
+            return FIRST;
+        }
+        if (matchCount == SECOND.matchCount) {
+            return getSecondOrThird(hasBonusNumber);
+        }
+        if (matchCount == FOURTH.matchCount) {
+            return FOURTH;
+        }
+        if (matchCount == FIFTH.matchCount) {
+            return FIFTH;
+        }
+        return NONE;
```
가급적 네이밍은 getXXX, setXXX 보다는 메서드의 행위를 드러낼 수 있는 네이밍을 고려해주세요. 🙏
코드를 짜고있는 나는 괴로울지라도 같이 일하는 동료들은 그 고민덕분에 코드를 더 쉽게 파악할 수 있을거에요. 😁

서비스가 발전해나감에따라 추후 Prize 요소가 더 추가되거나 혹은 반대로 더 줄어들 수도 있을 것 같은데요. 
지금의 getPrize는 요소가 추가되거나 삭제될 때마다 일일이 로직의 수정이 필요할 것 같아요. 😢 

enum 요소가 추가되더라도 getPrize 메서드를 수정하지 않아도 되는 방법은 없을까요?

kjyyjk replied:
일치 번호 개수와 보너스 번호 일치 여부에 따른 등수를 확인하는 것이기 때문에,
checkRank와 checkSecondOrThird로 메서드 명을 수정했습니다.

kjyyjk replied:
사실 로키가 지적해주신 부분에 대해 페어랑 많이 고민해보았었습니다.
처음 작성한 코드는 그렇지 않았는데,
프로그래밍 요구사항인 메서드 depth 문제를 해결하려다 보니 그런 코드로 변해간 거 같아요😢

우선 모든 요소에 대해 분기문을 가지지 않도록 로직을 수정했는데,
depth 문제(for문안에 if문)를 해결할 수 있는 좋은 방법이 있을까요?

Rok93 commented on src/main/java/lotto/Prize.java:
```diff
@@ -0,0 +1,49 @@
+        return NONE;
+    }
+
+    private static Prize getSecondOrThird(final boolean hasNumber) {
```
메서드 파라미터로 들어오는 값이 단지 number인가요? 🤔
실제로 메서드 파라미터로 들어오게되는 값은 보너스 볼일 것  같은데요. 
네이밍에서도 보너스볼이 들어오게될 것이라는 것을 드러내주면 좋을 것 같아요.

kjyyjk replied:
로키의 피드백처럼 hasNumber를 hasBonusNumber라고 명확하게 표현해주면 더 좋을 거 같아요!
수정했습니다.

Rok93 commented on src/main/java/lotto/WinningStatistics.java:
```diff
@@ -0,0 +1,28 @@
+    private int calculateTotal() {
+        int total = 0;
+        for (final Prize prize : winningStatistics.keySet()) {
+            total += winningStatistics.get(prize) * prize.getWinningAmount();
```
Prize [객체에게 메시지를 보내](https://tecoble.techcourse.co.kr/post/2020-04-28-ask-instead-of-getter/)보면 좋을 것 같아요. 😃

kjyyjk replied:
Rank에 당첨 개수를 파라미터로 받아 해당 등수의 총 당첨금을 반환하는 메서드를 만들고,
이에 대한 테스트도 추가했습니다!

Rok93 commented on src/test/java/lotto/LottoMachineTest.java:
```diff
@@ -0,0 +1,15 @@
+        Lotto result = LottoMachine.createLotto();
+        assertThat(result.getNumbers().size()).isEqualTo(6);
+    }
+}
```
PR 페이지에서 FileChanged로 코드를 볼 때, 지금처럼 빨간색의 `-` 표시가 보이는 경우가 있는데, 왜 보이는지 궁금하지 않으신가요? 😃

[참고자료](https://hyeon9mak.github.io/github-no-newline-at-a-end-of-file/)

kjyyjk replied:
파일 끝 개행처리..! 좋은 정보 알게되었습니다.
감사합니다🤗

Rok93 commented on src/test/java/lotto/PrizeTest.java:
```diff
@@ -0,0 +1,18 @@
+
+class PrizeTest {
+    @DisplayName("일치 개수와 보너스 번호 일치 여부에 따라서 Prize를 반환한다")
+    @CsvSource(value = {"6:false:FIRST", "5:true:SECOND", "5:false:THIRD", "4:false:FOURTH", "3:false:FIFTH"},
```
Parameterized Test 활용 👍💯 

일치하는 갯수가 5개를 제외한 케이스는 true, false 모두 동일한 값이 나와야할 것 같은데요. 
이 케이스에 대한 테스트 프로퍼티도 추가해주시면 좋을 것 같아요. 😃

kjyyjk replied:
Parameterized Test를 활용하니 더 간편하고 자세하게 테스트할 수 있어 좋은 거 같아요.
하나 궁금한 것이 있다면 실무에서도 Parameterized Test를 많이 사용하나요?
사실 모든 테스트에 Parametrized Test를 사용하라고 하면 사용할 수 있을 거 같은데,
주로 어떤 때 사용하는 것이 좋을지..
로키만의 기준이 궁금해요!

Rok93 replied:
크게 지금 레오가 작성한 이 테스트와 크게 다르지 않다고 생각해요. 😃
하나의 프로퍼티만이 아니라 다양한 프로퍼티에 대한 테스트가 필요한 경우. 
보통 Input과 Output만 딱 다른 경우에 많이 활용하게되는 것 같아요.

Rok93 reviewed (changes requested):
안녕하세요 레오. 로또 미션을 함께하게된 로키입니다.  잘 부탁드려요. 🙇‍♂️
전체적으로 고민의 흔적이 많이 보였던 것 같습니다. 👍👍 
질문에 대한 답변 그리고 추가로 코멘트 남겨두었으니 확인 부탁드릴게요. 😃

잘 이해가 안되거나 어려운 부분있으면 언제든지 DM 주세요. 🤗

kjyyjk committed:
style: 상수와 상태 사이 줄바꿈 추가

- 상수와 상태(클래스 필드)간에 줄바꿈을 해주어 가독성을 증진한다.

- 다른 클래스들에서는 상수와 생성자 간에 줄바꿈이 되어있으나,
OutputView는 그렇지 못하다. 따라서 OutputView도 통일시켜준다.

kjyyjk committed:
style: 등수를 Prize 대신 Rank로 표현

kjyyjk committed:
refactor: Rank의 getRank() 로직 수정

- 기존 checkRank()는 각 Prize 요소에 대한 분기문을 모두 작성하였다.
작성 당시에는 Prize의 요소가 적어 괜찮겠지하는 마음에 작성하였지만,
추후 Prize의 요소가 늘거나 줄게되면 일일이 메서드의 내부 로직도 수정해야한다.
따라서 이를 보완하는 방향으로 로직을 수정했다.

- 메서드의 의도를 드러내고자 getRank()에서 checkRank()로,
getSecondOrThird()에서 checkSecondOrThird()로 메서드 명을 수정하였다.

- checkSecondOrThird()의 두번째 파라미터는 보너스 볼 일치 여부이므로
hasBonusNumber로 변수명을 수정한다.

kjyyjk committed:
style: 파일 끝 개행 추가

- 파일 끝에 개행을 추가하지 않으면 Github에서는 이를 불완전하다고 보고
경고를 보내준다. 따라서 파일 끝 개행을 추가한다.

- 매번 체크하기 힘들 수 있기 때문에 Intellij Editor에 개행 보장기능을 설정했다.

kjyyjk committed:
refactor: Rank가 당첨 개수를 받아 당첨금을 계산

kjyyjk committed:
docs: 기능 목록에서 예외 케이스를 관리

kjyyjk committed:
refactor: WinningLotto가 당첨 번호와 보너스 번호를 관리

- WinningLotto 내부에서 당첨 번호와 보너스 번호를 관리한다.
따라서 기존에 Application 내에 존재했던 당첨-보너스 중복 검증 로직을
WinningLotto 내부로 이동한다.
또한 기존에 Application 내에 존재했던 당첨 번호 검증 로직을
WinningNumbers 내부로 이동한다.

- 그에 따른 WinningLotto 테스트와 WinningNumbers 테스트를 수행한다.

kjyyjk committed:
refactor: 당첨통계 계산 로직을 WinningLotto에 위임

Rok93 commented on src/main/java/lotto/LottoManager.java:
```diff
@@ -0,0 +1,36 @@
+        }
+
+        if (purchaseAmount <= 0) {
+            throw new IllegalArgumentException("구입금액은 양수여야 합니다.");
```
입력값이 무엇인지 예외 메시지에서 보여준다면, 에러를 잡기가 조금 더 수월해질 것 같아요. 😉

Rok93 commented on src/main/java/lotto/Rank.java:
```diff
@@ -0,0 +1,49 @@
+package lotto;
+
+public enum Rank {
+    NONE(0, 0),
+    FIFTH(5000, 3),
+    FOURTH(50000, 4),
+    THIRD(1500000, 5),
+    SECOND(30000000, 5),
+    FIRST(2000000000, 6);
```
숫자의 경우 3자리 단위로 `_`를 추가해줄 수 있는데요. 
이렇게 `_`를 넣어주면 숫자 값을 파악하기 수월할 것 같네요. 😉 

```suggestion
    NONE(0, 0),
    FIFTH(5_000, 3),
    FOURTH(50_000, 4),
    THIRD(1_500_000, 5),
    SECOND(30_000_000, 5),
    FIRST(2_000_000_000, 6);
```

Rok93 commented on src/main/java/lotto/Rank.java:
```diff
@@ -0,0 +1,49 @@
+    }
+
+    public int calculateAllWinningAmount(final int winningCount) {
+        return winningCount * getWinningAmount();
```
이 경우에는 객체 자신의 상태 값을 사용하는거라면 굳이 getter로 값을 가져올 필요는 없을 것 같아요. 😃

Rok93 commented on src/main/java/lotto/Rank.java:
```diff
@@ -0,0 +1,49 @@
+    }
+
+    public static Rank checkRank(final int matchCount, final boolean hasBonusNumber) {
+        for (final Rank rank : Rank.values()) {
+            if (matchCount == SECOND.matchCount) {
+                return checkSecondOrThird(hasBonusNumber);
+            }
+            if (rank.matchCount == matchCount) {
+                return rank;
+            }
+        }
+        return NONE;
```
아래와 같이 변경해볼 수도 있을 것 같네요. 🤗 
```suggestion
        if (matchCount == SECOND.matchCount) {
            return checkSecondOrThird(hasBonusNumber);
        }

        return Arrays.stream(Rank.values())
            .filter(rank -> rank.matchCount == matchCount)
            .findAny()
            .orElse(NONE);
```

Rok93 replied:
> (질문) 사실 로키가 지적해주신 부분에 대해 페어랑 많이 고민해보았었습니다.
처음 작성한 코드는 그렇지 않았는데,
프로그래밍 요구사항인 메서드 depth 문제를 해결하려다 보니 그런 코드로 변해간 거 같아요😢
우선 모든 요소에 대해 분기문을 가지지 않도록 로직을 수정했는데,
depth 문제(for문안에 if문)를 해결할 수 있는 좋은 방법이 있을까요?

상황에 따라 다르기 때문에 딱 잘라서 말하기는 어렵지만 
결국에 depth를 줄이기위한 가장 쉬운 접근법은 메서드를 추출하는 것이라 생각해요. 😃
답변이 제가 말하고도 조금 아쉬운 것 같은데... 혹시 depth 줄이기에 고민되는 상황이 발생하면 공유주세요. 😁

Rok93 commented on src/main/java/lotto/WinningNumbers.java:
```diff
@@ -0,0 +1,34 @@
+        }
+
+        for (final int winningNumber : winningNumbers) {
+            Lotto.validateLottoNumber(winningNumber);
```
잘 생각해보면 WinningNumbers도 Lotto와 동일한 검증이 필요하고 생각해보면 Lotto도 WinningNumbers처럼 6자리의 길이 제한이 있고 (물론 Lotto에서는 검증로직이 없군요 👀)

WinningLottos와 Lotto는 동일한 것으로 볼 수 있지 않을까요? 🤔

Rok93 commented on src/main/java/lotto/WinningStatistics.java:
```diff
@@ -0,0 +1,28 @@
+    }
+
+    private int calculateTotalWinningAmount() {
+        int total = 0;
+        for (final Rank rank : winningStatistics.keySet()) {
+            total += rank.calculateAllWinningAmount(winningStatistics.get(rank));
+        }
+        return total;
```
아래와 같이 변경해볼 수도 있을 것 같아요. 😉 

```suggestion
        return winningStatistics.entrySet()
            .stream()
            .mapToInt(entry -> entry.getKey().calculateAllWinningAmount(entry.getValue()))
            .sum();
```

Rok93 commented on src/main/java/lotto/WinningLotto.java:
```diff
@@ -0,0 +1,50 @@
+    private int calculateMatchCount(final List<Integer> lottoNumbers, final List<Integer> winningNumbers) {
+        List<Integer> matchNumbers = new ArrayList<>(winningNumbers);
+        matchNumbers.retainAll(lottoNumbers);
+        int matchCount = matchNumbers.size();
+        return matchCount;
```
불필요한 변수 선언을 제거해줄 수 있을 것 같네요. 
```suggestion
        return matchNumbers.size();
```

Rok93 commented on src/test/java/lotto/LottoTest.java:
```diff
@@ -0,0 +1,19 @@
+    @DisplayName("로또는 1과 45 사이의 번호가 아니면 예외를 던진다")
+    @Test
+    void 로또는_1과_45_사이의_번호가_아니면_예외를_던진다() {
+        assertThatThrownBy(() -> validateLottoNumber(46))
+                .isInstanceOf(IllegalArgumentException.class)
+                .hasMessage("로또 번호는 %d ~ %d 사이여야 합니다.".formatted(MIN_LOTTO_NUMBER, MAX_LOTTO_NUMBER));
```
validateLottoNumber 메서드를 테스트하기보다는 
Lotto 객체를 생성할 때, 예외가 적절히 발생하는지를 검증하는 테스트를 해주시면 좋을 것 같습니다. 😃

Rok93 commented on src/test/java/lotto/LottoMachineTest.java:
```diff
@@ -0,0 +1,15 @@
+    @Test
+    void 로또는_6개의_번호를_가진다() {
+        Lotto result = LottoMachine.createLotto();
+        assertThat(result.getNumbers().size()).isEqualTo(6);
```
각 로또 숫자 하나하나가 1 ~ 45 사이의 숫자인지도 확인하는 검증문이 있으면 좋을 것 같아요.

Rok93 commented on src/main/java/lotto/WinningLotto.java:
```diff
@@ -0,0 +1,50 @@
+        this.bonusNumber = bonusNumber;
+    }
+
+    public WinningStatistics calculateStatistics(final List<Lotto> lottos) {
```
해당 메서드에 대한 테스트가 누락된 것 같아요. 👀
테스트 코드 추가 부탁드려요. 😉

Rok93 commented on src/main/java/lotto/WinningNumbers.java:
```diff
@@ -0,0 +1,34 @@
+                .toList();
+    }
+
+    public boolean contains(final int number) {
```
contain 메서드에 대한 테스트도 추가 부탁드릴게요.

Rok93 commented on src/main/java/lotto/WinningStatistics.java:
```diff
@@ -0,0 +1,28 @@
+        this.winningStatistics = winningStatistics;
+    }
+
+    public double calculateReturnRate(final int purchaseAmount) {
```
이 메서드도 테스트 코드 추가 부탁드려요. 😉

Rok93 reviewed (approved):
레오 피드백 반영 잘해주셨습니다. 👍💯
1단계는 이정도면 충분한 것 같아서 이만 머지하도록하겠습니다. 👏🚀
몇몇 코멘트 남겨두었는데, 다음 단계 진행하면서 같이 확인 부탁드릴게요. 😉 
2단계에서 기다리고잇겠습니다. 🤗
